### ISSUE TITLE
Why `from` is more sound than `toBuilder`

### ISSUE THREAD

**Issue by andrey-vasilyev** [immutables/immutables#1195]
Could someone please elaborate on the following statement from the documentation? 

`The from method on builders is a much more sound and powerful alternative than having toBuilder() (or alike) on immutable values.`

What are the pros and cons? How is `age.toBuilder().age(42).build()` is really different to `ImmutableAge.withAge(24)`? Both start on immutable value! Second is shorter of course, but that is not the point.

I find it less powerful actually. Because if I have builders and private implementations, I have to do `new AgeBuilder().from(age)`, while `age.toBuilder()` is shorter and could be called by method reference `Age::toBuilder` - you can't do both with `from` approach.

**Comment by elucash** [comment]
These are all opinions, make you own choice. For me, at least initially, it was beneficial to not pollute data value interface with things generated from it, so you can have clean abstract value type `X`, and generated `ImmutableX`, 'ModifiableX' etc (akin to `List`, `ImmutableList`, `ArrayList`, etc). Of course, since we have many more styles of usages, including "sandwich" `interface X {  int attr(); class Builder extends ImmutableX.Builder {}}..... new X.Builder().attr(1).build()`, which makes initial principles less relevant, but anyway. With methods are different because they try to create copies with structural sharing, builder would produce a lot more allocations for non-trivial objects, so I consider it not the same tool as builder. But `Builder.from` is more or less what `toBuilder` is doing. What from is good for is to collect contents of multiple source object into resulting, also it supports filling properties from supertypes of abstract value types, not only the value type itself. Given that `Builder.from` and its overloads are so powerful, then why bother having `toBuilder` too?
The curse of modern software/programming languages when literally the same can be trivially done in many ways and you have to waste brain cycles on deciding which variation to choose (`Instant.now(clock)` or `clock.instant()`, `toBuilder` or `Builder.from`, foreach loop or forEach method..)

**Comment by andrey-vasilyev** [comment]
Thanks for the answer. I get it. But if I still prefer `toBuilder` (I don't need collecting properties from multiple abstract classes as I intend to have only one and hide immutable value impl) is there a way to generate it instead of `from`? Or the only way is to implement it myself every time?

**Comment by elucash** [comment]
There no such feature now, we might implement it, as for example opt-in naming style: we have some style flags / namings with empty defaults, so that once you give a name to a method in styles it enables it to be generated, but by default it is not generated, as an example: `Value.Style.canBuild`, but in interface you'll still need to create `abstract Builder toBuilder();`. Also for not this new feature seems to be not priority, unless some more people would want it (and it's mostly on backburner for now, we're making some releases with minor fixes / refinements from time to time). Creating such method manually would a bit longer, for sure `default Builder toBuilder() { return ImmutableX.builder().from(this); }`, but not that bad.

**Comment by andrey-vasilyev** [comment]
Ok, thanks. If I try an contribute this feature, would you accept it?

**Comment by elucash** [comment]
yes, PR welcome (yep, we a little be behind on current PRs but we can go over them soon). I just think that there could be some nuances in implementation, like generated toBuilder should probably just call `Builder.from(this)`, but `from` method is generated if there are no conflicts with `from` attributes and not `strictBuilder` etc etc. so implementing it kinda easy, but not trivial, but I can help if you stumble on this.

**Comment by andrey-vasilyev** [comment]
Cool. Actually, I am not sure what would be a best way to configure that behavior. Like check if the abstract class has `toBuilder()` method and then immutables should generate it? 

And this won't be fast (if it ever happens) I've never coded an annotation processor before ;)
